x-pack/plugin/core/src/test/java/org/elasticsearch/protocol/AbstractHlrcStreamableXContentTestCase.java

```diff
@@ -43,4 +43,10 @@ public final void testHlrcFromXContent() throws IOException {
      */
     public abstract T convertHlrcToInternal(H instance);
 
+    //TODO this would be final ideally: why do both responses need to parse from xcontent, only one (H) should? I think that T#fromXContent
+    //are only there for testing and could go away?
+    @Override
+    protected T doParseInstance(XContentParser parser) throws IOException {
+        return convertHlrcToInternal(doHlrcParseInstance(parser));
+    }
```

Comment:
@vladimirdolzhenko thoughts?

Reply:
in this case you hide xpack-parsing-to-xpack-instance functionality that I don't want to  drop.

Reply:
I see your point.

Usually we have both `toXContent` and `fromXContent` in xpack request/response and `fromXContent` only on HLRC side. I'd like to keep `fromXContent` on a server side (maybe another clean up iteration is required)

Reply:
anyway - I would not go with this change on a level of `AbstractHlrcStreamableXContentTestCase`

Reply:
I think this is a mistake. To me, the main point of having such a test class under xpack is to avoid code duplication. Otherwise it would be easy to have tests under client that rely on e.g. toXContent pasted from xpack that's not needed in client. Given where tests currently are (under x-pack), code duplication is not required, but we end up duplicating it under x-pack and client. I don't think we should maintain this moving forward. Why keep functionality that is not needed?

This change in my PR does not affect existing tests that override this method, as the method is not yet made final. I think though that this is a better default behaviour than asking people to duplicate fromXContent, less code to maintain, and less hassle in writing tests.

Reply:
> I'd like to keep fromXContent on a server side (maybe another clean up iteration is required)

How can we do that? Response parsing is needed in the client only, which should not depend on server (at least ideally).

Reply:
Agree - it's a big step forward. But we have to clean up many other leftovers in this case. @hub-cap @nik9000 your objections ?

Reply:
Maybe we're extending the wrong test class here. That is why I built that fluent testing thingy - I figured we'd want to be fairly surgical in what we tested. I don't think we should maintain response parsing in the server. If we can get away with dropping it. I don't like keeping it just for testing.

Reply:
maybe the problem here, and where the confusion comes from, is that requests need parsing in server, and toXContent in client, while responses need parsing in client and toXContent in server. I guess we can't have a single test class that works well for both requests and responses? And my change here is only addressing responses although the test is meant to be more generic though it is used at the moment only for responses.